src/crc32c.c

```diff
@@ -98,16 +101,14 @@ static uint32_t crc32c_sse42(uint32_t crc, const char *buf, size_t len)
        to an eight-byte boundary */
     while (len && ((uintptr_t)buf & 7) != 0) {
         __asm__("crc32b\t" "(%1), %0"
-                : "=r"(crc0)
-                : "r"(buf), "0"(crc0));
+                : "+r"(crc0)
+                : "r"(buf), "m"(*buf));
```

Comment:
If I understand my asm constraint parsing correctly at https://gcc.gnu.org/onlinedocs/gcc/Extended-Asm.html#Input-Operands, you have changed this from a constraint that it must pick a single register for output (`=r`) that is also chosen as an input register (`r`) with the output assigned to the variable crc0 and the input assigned from the variable crc0 to a constraint that is must pick a single register for input and output (`+r`) for the variable crc0 (this change should not affect code generation)

Secondly, you have added a constraint that the `asm` block will read one byte as represented by the memory expression `*buf` (other C expressions are valid here, such as `buf[0]` to show it is an array read or `((uint64_t*)buf)[0]` to show it is an aligned 64-bit read or `(*(const char (*)[10]) buf)` to show it reads the 10 bytes from `buf[0]` to `buf[9]`

Reply:
Note that the readable/alignment constraints are subtractive, not additive. They prevent the compiler from making optimizations that would eliminate those properties, but they do not create constraints that the compiler needs to meet those properties (for TBAA, alignment, etc). Required constraints must already be met during declaration/allocation, with the right type and alignment, for example.

Reply:
> Secondly, you have added a constraint that the asm block will read one byte as represented by the memory expression `*buf`

Based on the SO comment above, this is a dummy argument that implicitly fixes the safety of the preceding `"r"(buf)` argument somehow.

Reply:
Right, the `"r"(buf)` constraint (unchanged) says that the variable `buf` is assigned to the expression `%0` which must be an input register, but says nothing about it being a pointer. The `"m"(*buf)` says that the value `*buf` must be computed by the expression `%3` (which is unused in the actual assembly) and must be valid as input

Reply:
Another comment on SO also noted similarly, for that reason, that these could be combined as something like
```
__asm__ ("crc32b\t" "%1, %0"
                : "=r"(crc0)
                : "rm"(*buf) : )
```

which instead says that `%1` must be either the value `*buf` in a register or as a memory load expression that computes  `*buf`

Reply:
Thanks!

I guess I'll leave it as-is, since in either case it sounds like code generation is unchanged, and in the current form it is easier to compare to the upstream version.  But `"rm"(*buf)` does look a bit cleaner.

Reply:
the actual number of bytes being accessed by `CRC32_PTR` is `sizeof(void*)`, no?